qiskit/circuit/quantumcircuit.py

```diff
@@ -4886,28 +4863,100 @@ def _compare_parameters(param1: Parameter, param2: Parameter) -> int:
+
+
+def _qasm2_define_custom_operation(operation, existing_gate_names, gates_to_define):
+    """Extract a custom definition from the given operation, and append any necessary additional
+    subcomponents' definitions to the ``gates_to_define`` ordered dictionary.
+
+    Returns a potentially new :class:`.Instruction`, which should be used for the
+    :meth:`~.Instruction.qasm` call (it may have been renamed)."""
+    from qiskit.circuit import library as lib  # pylint: disable=cyclic-import
+
+    if operation.name in existing_gate_names:
+        return operation
+
+    # Check instructions names or label are valid
+    escaped = _qasm_escape_name(operation.name, "gate_")
+    if escaped != operation.name:
+        operation = operation.copy(name=escaped)
+
+    # These are built-in gates that are known to be safe to construct by passing the correct number
+    # of `Parameter` instances positionally, and have no other information.  We can't guarantee that
+    # if they've been subclassed, though.  This is a total hack; ideally we'd be able to inspect the
+    # "calling" signatures of Qiskit `Gate` objects to know whether they're safe to re-parameterise.
+    known_good_parameterized = {
+        lib.PhaseGate,
+        lib.RGate,
+        lib.RXGate,
+        lib.RXXGate,
+        lib.RYGate,
+        lib.RYYGate,
+        lib.RZGate,
+        lib.RZXGate,
+        lib.RZZGate,
+        lib.XXMinusYYGate,
+        lib.XXPlusYYGate,
+        lib.UGate,
+        lib.U1Gate,
+        lib.U2Gate,
+        lib.U3Gate,
+    }
+
+    # In known-good situations we want to use a manually parametrised object as the source of the
+    # definition, but still continue to return the given object as the call-site object.
+    if type(operation) in known_good_parameterized:
+        parameterized_operation = type(operation)(*_QASM2_FIXED_PARAMETERS[: len(operation.params)])
+    elif hasattr(operation, "_qasm2_decomposition"):
+        new_op = operation._qasm2_decomposition()
+        parameterized_operation = operation = new_op.copy(
+            name=_qasm_escape_name(new_op.name, "gate_")
+        )
+    else:
+        parameterized_operation = operation
+
+    # If there's an _equal_ operation in the existing circuits to be defined, then our job is done.
+    previous_definition_source, _ = gates_to_define.get(operation.name, (None, None))
+    if parameterized_operation == previous_definition_source:
+        return operation
+
+    # Otherwise, if there's a naming clash, we need a unique name.
+    if operation.name in gates_to_define:
+        new_name = f"{operation.name}_{id(operation)}"
+        operation = operation.copy(name=new_name)
+    else:
+        new_name = operation.name
+
+    if parameterized_operation.params:
+        parameters_qasm = (
+            "(" + ",".join(f"param{i}" for i, _ in enumerate(parameterized_operation.params)) + ")"
+        )
```

Comment:
```suggestion
            "(" + ",".join(f"param{i}" for i in range(len(parameterized_operation.params))) + ")"
        )
```

Reply:
Fine by me.  I think both are reasonably idiomatic, and the `range` form involves less allocation work.